src/actions/apply_attack_action.rs

```diff
@@ -76,6 +76,64 @@ fn binomial_coefficient(n: usize, k: usize) -> usize {
     result
 }
 
+fn gholdengo_scintillating_surfing(
+    acting_player: usize,
+    state: &State,
+) -> (Probabilities, Mutations) {
+    let active_pokemon = state.get_active(acting_player);
+    let metal_energy_count = active_pokemon
+        .attached_energy
+        .iter()
+        .filter(|&energy| *energy == EnergyType::Metal)
+        .count();
+
+    if metal_energy_count == 0 {
+        // No metal energy attached, no coins to flip
+        return probabilistic_damage_attack(vec![1.0], vec![0]);
+    }
+
+    // Generate all possible outcomes for flipping N coins
+    let num_outcomes = 2_usize.pow(metal_energy_count as u32);
+    let mut probabilities = vec![0.0; metal_energy_count + 1]; // 0 to metal_energy_count heads
+    let mut damages = Vec::new();
+
+    // For each possible outcome (0 to metal_energy_count heads)
+    for (heads, prob) in probabilities.iter_mut().enumerate() {
+        // Probability of getting exactly 'heads' heads out of 'metal_energy_count' coins
+        *prob = binomial_coefficient(metal_energy_count, heads) as f64 / (num_outcomes as f64);
+        damages.push((heads as u32) * 50); // 50 damage per heads
+    }
+
+    probabilistic_damage_attack(probabilities, damages)
+}
```

Comment:
Can you refactor this to re-use most of the code with the Celebi one? Maybe one re-usable function that optionally accepts a type of energy to filter by? If none specify, any energy counts towards the amount of coin flips.